src/node-realm.ts

```diff
@@ -1,46 +1,58 @@
-import { SecureEnvironment } from "./environment";
-import { RedProxyTarget, BlueFunction } from "./types";
-import { getOwnPropertyDescriptors, construct, ErrorCreate } from "./shared";
+import { MembraneBroker } from "./environment";
+import { RedProxyTarget } from "./types";
+import { assign, isUndefined } from "./shared";
 import { runInNewContext } from 'vm';
+import { blueProxyFactory, controlledEvaluator } from "./blue";
+import { getFilteredEndowmentDescriptors, linkIntrinsics } from "./intrinsics";
+import { SandboxRegistry } from "./registry";
+import { serializedRedEnvSourceText } from "./red";
 
 // note: in a node module, the top-level 'this' is not the global object
 // (it's *something* but we aren't sure what), however an indirect eval of
 // 'this' will be the correct global object.
 const unsafeGlobalEvalSrc = `(0, eval)("'use strict'; this")`;
 
-export default function createSecureEnvironment(distortionMap?: Map<RedProxyTarget, RedProxyTarget>, endowments?: object): (sourceText: string) => void {
+interface NodeEvaluationOptions {
+    // the blue global object to be used to create the sandbox
+    globalThis: typeof globalThis;
+    // additional globals to be installed inside the sandbox
+    endowments: object;
+    // distortions map
+    distortions?: Map<RedProxyTarget, RedProxyTarget>;
+}
+
+export default function createSandboxEvaluator(registry: SandboxRegistry, options: NodeEvaluationOptions): (sourceText: string) => void {
     // Use unsafeGlobalEvalSrc to ensure we get the right 'this'.
-    const redGlobalThis = runInNewContext(unsafeGlobalEvalSrc);
+    const redGlobalThis: typeof globalThis = runInNewContext(unsafeGlobalEvalSrc);
+    const { globalThis: blueGlobalThis, endowments } = options;
+    const endowmentsDescriptors = getFilteredEndowmentDescriptors(endowments);
     const { eval: redIndirectEval } = redGlobalThis;
-    const blueGlobalThis = globalThis as any;
-    const env = new SecureEnvironment({
-        blueGlobalThis,
-        redGlobalThis,
-        distortionMap,
-    });
+    linkIntrinsics(registry, blueGlobalThis, redGlobalThis);
+
+    const redProxyFactory = redIndirectEval(`(${serializedRedEnvSourceText})`);
+
+    const broker = new MembraneBroker(
+        registry,
+        (broker: MembraneBroker) => blueProxyFactory(broker, redGlobalThis.Reflect),
+        (broker: MembraneBroker) => redProxyFactory(broker, blueGlobalThis.Reflect)
+    );
 
     // remapping globals
-    env.remap(redGlobalThis, blueGlobalThis, endowments && getOwnPropertyDescriptors(endowments));
-
-    return (sourceText: string): void => {
```

Comment:
this logic was shared between node and browsers, we abstract it out into `controlledEvaluator`